File: packages/mobile/native/android/java/com/bendyline/gezel/mobile/GezelSpeechPlugin.java
package com.bendyline.gezel.mobile;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;
import android.os.PowerManager;
import android.os.SystemClock;
import android.util.Base64;
import com.getcapacitor.*;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.genai.common.FeatureStatus;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.*;
import org.json.*;

@CapacitorPlugin(name = "GezelSpeech")
public final class GezelSpeechPlugin extends Plugin {
    private final ExecutorService worker = Executors.newSingleThreadExecutor();
    private final ExecutorService probes = Executors.newSingleThreadExecutor();
    private SpeechAssets assets;
    private String activeId;
    private boolean backgrounded, stopped;
    private long engine;
    private MlKitSpeech system;
    private final List<PluginCall> waiters = new ArrayList<>();
    @Override public void load() { assets = new SpeechAssets(getContext()); }
    @PluginMethod public void status(PluginCall call) {
        probes.execute(() -> {
            JSObject systemStatus = new JSObject().put("state", "unavailable");
            if (Build.VERSION.SDK_INT >= 31) {
                try (MlKitSpeech probe = new MlKitSpeech(call.getString("language"))) {
                    int status = probe.status();
                    systemStatus.put("state", status == FeatureStatus.AVAILABLE ? "ready" : status == FeatureStatus.DOWNLOADABLE || status == FeatureStatus.DOWNLOADING ? "download-required" : "unavailable");
                    systemStatus.put("model", "system");
                } catch (Exception | LinkageError unavailable) { systemStatus.put("reason", "On-device speech is unavailable; Gezel can use Whisper offline."); }
            }
            try {
                String state = assets.bundled() ? "ready" : "unavailable";
                call.resolve(new JSObject().put("system", systemStatus)
                    .put("whisper", new JSObject().put("state", state).put("model", "whisper-tiny"))
                    .put("kokoro", new JSObject().put("state", state).put("model", "kokoro-82m-v1.0"))
                    .put("models", assets.models())
                    .put("voices", assets.bundled() ? assets.voices() : new JSONArray()));
            } catch (Exception error) { call.reject("The offline speech pack could not be read", "unavailable"); }
        });
    }
    @PluginMethod public void transcribe(PluginCall call) { run(call, false); }
    @PluginMethod public void synthesize(PluginCall call) { run(call, true); }
    private void run(PluginCall call, boolean synthesis) {
        String id = call.getString("requestId");
        try { UUID.fromString(id); } catch (Exception invalid) { call.reject("A speech request identity is required", "invalid-input"); return; }
        synchronized (this) {
            if (activeId != null || backgrounded) { call.reject("Speech is busy or Gezel is in the background", "busy"); return; }
            activeId = id; stopped = false;
        }
        worker.execute(() -> {
            JSObject result = null; String failure = null, code = "failed";
            long start = SystemClock.uptimeMillis();
            try {
                check();
                if (synthesis) {
                    String model = call.getString("model", "kokoro-82m-v1.0");
                    if (!model.equals("kokoro-82m-v1.0")) throw new IllegalArgumentException("The selected Kokoro model is unavailable");
                    String text = call.getString("text", ""), voice = call.getString("voice", "af_heart");
                    if (text.isEmpty() || text.length() > 12000) throw new IllegalArgumentException("Use up to 12,000 characters of speech");
                    int index = -1; JSONArray voices = assets.voices();
                    for (int i = 0; i < voices.length(); i++) if (voices.getJSONObject(i).getString("id").equals(voice)) index = voices.getJSONObject(i).getInt("index");
                    if (index < 0) throw new IllegalArgumentException("The selected Kokoro voice is unavailable");
                    File root = assets.ensure(); checkResources(); check();
                    long nativeEngine = allocate();
                    byte[] wav = SpeechRuntime.synthesize(nativeEngine, new File(root, "kokoro").getPath(), text, index, (float) Math.min(2, Math.max(.5, call.getDouble("speed", 1.0))));
                    check();
                    result = new JSObject().put("wav", Base64.encodeToString(wav, Base64.NO_WRAP)).put("meta", new JSObject()
                        .put("voice", voice).put("model", model).put("sampleRate", 24000).put("durationSeconds", (wav.length - 44) / 48000.0).put("durationMs", SystemClock.uptimeMillis() - start));
                } else {
                    String encoded = call.getString("audio", "");
                    if (encoded.isEmpty() || encoded.length() > 5120000) throw new IllegalArgumentException("Use up to two minutes of speech");
                    byte[] pcm = Base64.decode(encoded, Base64.NO_WRAP);
                    if (pcm.length % 2 != 0 || pcm.length == 0 || !Base64.encodeToString(pcm, Base64.NO_WRAP).equals(encoded)) throw new IllegalArgumentException("Invalid speech recording");
                    String provider = call.getString("engine", "");
                    String text;
                    if (provider.equals("system")) {
                        if (Build.VERSION.SDK_INT < 31) throw new MlKitSpeech.Unavailable();
                        synchronized (this) { check(); system = new MlKitSpeech(call.getString("language")); }
                        text = system.transcribe(pcm);
                    } else if (provider.equals("whisper")) {
                        if (!call.getString("model", "whisper-tiny").equals("whisper-tiny")) throw new IllegalArgumentException("The selected Whisper model is unavailable");
                        File root = assets.ensure(); checkResources(); check();
                        text = new String(SpeechRuntime.transcribe(allocate(), new File(root, "whisper-tiny.bin").getPath(), pcm, call.getString("language"), call.getString("prompt")), StandardCharsets.UTF_8);
                    } else throw new IllegalArgumentException("Unknown speech recognizer");
                    check(); result = new JSObject().put("text", text).put("durationMs", SystemClock.uptimeMillis() - start);
                }
            } catch (MlKitSpeech.Unavailable unavailable) { failure = "On-device speech is unavailable"; code = "unavailable"; }
            catch (CancellationException cancelled) { failure = "Speech stopped"; code = "cancelled"; }
            catch (Exception | LinkageError | OutOfMemoryError error) { failure = error instanceof OutOfMemoryError ? "Not enough memory for speech" : error.getMessage(); }
            finally {
                MlKitSpeech previous; synchronized (this) { previous = system; system = null; }
                if (previous != null) previous.close();
                synchronized (this) {
                    if (engine != 0) SpeechRuntime.destroy(engine); engine = 0;
                    if (stopped) { failure = "Speech stopped"; code = "cancelled"; }
                    activeId = null;
                    if (failure == null && result != null) call.resolve(result); else call.reject(failure == null ? "Speech failed" : failure, code);
                    for (PluginCall waiter : waiters) waiter.resolve(); waiters.clear();
                }
            }
        });
    }
    private synchronized long allocate() {
        check(); engine = SpeechRuntime.create();
        if (engine == 0) throw new IllegalStateException("Not enough memory for speech");
        return engine;
    }
    private synchronized void check() { if (stopped || backgrounded) throw new CancellationException(); }
    private void checkResources() {
        ActivityManager manager = (ActivityManager) getContext().getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo memory = new ActivityManager.MemoryInfo(); manager.getMemoryInfo(memory);
        if (memory.lowMemory || memory.availMem < memory.threshold + 512L * 1024 * 1024) throw new IllegalStateException("Not enough memory for speech. Release the chat model and try again.");
        if (Build.VERSION.SDK_INT >= 29 && ((PowerManager) getContext().getSystemService(Context.POWER_SERVICE)).getCurrentThermalStatus() >= PowerManager.THERMAL_STATUS_SEVERE)
            throw new IllegalStateException("Let this device cool before running speech");
    }
    private synchronized void stop() {
        stopped = true;
        if (engine != 0) SpeechRuntime.cancel(engine);
        if (system != null) system.cancel();
    }
    @PluginMethod public synchronized void cancel(PluginCall call) {
        if (activeId != null && activeId.equals(call.getString("requestId"))) { waiters.add(call); stop(); }
        else call.resolve();
    }
    @Override protected synchronized void handleOnStop() { backgrounded = true; stop(); }
    @Override protected synchronized void handleOnStart() { backgrounded = false; }
    @Override protected synchronized void handleOnDestroy() { backgrounded = true; stop(); worker.shutdown(); probes.shutdown(); }
}

File: packages/mobile/native/ios/Plugin/GezelSpeechPlugin.swift
import Capacitor
import Foundation
import Speech
import UIKit
import AVFoundation
import GezelSpeech

@objc(GezelSpeechPlugin)
public final class GezelSpeechPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GezelSpeechPlugin"
    public let jsName = "GezelSpeech"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "status", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "transcribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "synthesize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise)
    ]
    private var active: CAPPluginCall?
    private var recognizer: SFSpeechRecognizer?
    private var task: SFSpeechRecognitionTask?
    private var backgroundObserver: NSObjectProtocol?
    private let worker = DispatchQueue(label: "com.bendyline.gezel.speech", qos: .userInitiated)
    private var nativeEngine: OpaquePointer?
    private var nativeCancelled = false
    private var cancelWaiters: [CAPPluginCall] = []
    private var speechRoot: URL { Bundle.main.bundleURL.appendingPathComponent("speech") }
    private var voices: [[String: Any]] {
        guard let data = try? Data(contentsOf: speechRoot.appendingPathComponent("voices.json")),
              let values = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return [] }
        return values
    }
    private func ready(_ name: String) -> Bool { FileManager.default.fileExists(atPath: speechRoot.appendingPathComponent(name).path) }
    private var models: [String: Any] {
        guard let data = try? Data(contentsOf: speechRoot.appendingPathComponent("pack.json")),
              let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return ["stt": [], "tts": []] }
        return value
    }

    public override func load() {
        backgroundObserver = NotificationCenter.default.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main) { [weak self] _ in
            self?.finish(error: "Speech stopped when Gezel entered the background", code: "cancelled")
        }
    }
    deinit {
        if let backgroundObserver { NotificationCenter.default.removeObserver(backgroundObserver) }
        task?.cancel()
    }
    private func locale(_ call: CAPPluginCall) -> Locale {
        call.getString("language").map { Locale(identifier: $0) } ?? Locale.current
    }
    private func systemStatus(_ locale: Locale) -> [String: Any] {
        guard let recognizer = SFSpeechRecognizer(locale: locale), recognizer.supportsOnDeviceRecognition else {
            return ["state": "unavailable", "reason": "On-device recognition is unavailable for this language."]
        }
        let permission = SFSpeechRecognizer.authorizationStatus()
        if permission != .authorized {
            return ["state": "permission-required", "language": locale.identifier, "model": "system", "reason": "Allow speech recognition to transcribe on this device."]
        }
        return ["state": recognizer.isAvailable ? "ready" : "unavailable", "language": locale.identifier, "model": "system"]
    }
    @objc public func status(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "system": self.systemStatus(self.locale(call)),
                "whisper": ["state": self.ready("whisper-tiny.bin") ? "ready" : "unavailable", "model": "whisper-tiny"],
                "kokoro": ["state": self.ready("kokoro/model.int8.onnx") ? "ready" : "unavailable", "model": "kokoro-82m-v1.0"],
                "voices": self.voices, "models": self.models
            ])
        }
    }
    @objc public func transcribe(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard self.active == nil, UIApplication.shared.applicationState != .background else {
                call.reject("Speech is busy or Gezel is in the background", "busy"); return
            }
            guard call.getString("requestId").flatMap(UUID.init(uuidString:)) != nil else {
                call.reject("A valid speech request identity is required", "invalid-input"); return
            }
            if call.getString("engine") == "whisper" {
                self.runNative(call, synthesis: false); return
            }
            guard call.getString("engine") == "system" else {
                call.reject("Unknown speech recognizer", "invalid-input"); return
            }
            guard let recognizer = SFSpeechRecognizer(locale: self.locale(call)), recognizer.supportsOnDeviceRecognition else {
                call.reject("On-device recognition is unavailable for this language", "unavailable"); return
            }
            self.active = call
            self.recognizer = recognizer
            let begin = { [weak self] in
                guard let self, self.active === call else { return }
                guard SFSpeechRecognizer.authorizationStatus() == .authorized else {
                    self.finish(error: "Allow speech recognition in Settings to use on-device transcription", code: "permission-required"); return
                }
                self.recognize(call, recognizer)
            }
            if SFSpeechRecognizer.authorizationStatus() == .notDetermined {
                SFSpeechRecognizer.requestAuthorization { _ in DispatchQueue.main.async(execute: begin) }
            } else { begin() }
        }
    }
    private func recognize(_ call: CAPPluginCall, _ recognizer: SFSpeechRecognizer) {
        guard let encoded = call.getString("audio"), encoded.count <= 5_120_000,
                  let data = Data(base64Encoded: encoded), !data.isEmpty, data.count <= 3_840_000, data.count % 2 == 0,
                  data.base64EncodedString() == encoded else {
                finish(error: "Use up to two minutes of mono 16 kHz speech", code: "invalid-input"); return
            }
            guard let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: false),
                  let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(data.count / 2)),
                  let samples = buffer.int16ChannelData else { finish(error: "Audio allocation failed"); return }
            buffer.frameLength = AVAudioFrameCount(data.count / 2)
            data.copyBytes(to: UnsafeMutableRawBufferPointer(start: samples[0], count: data.count))
            let request = SFSpeechAudioBufferRecognitionRequest()
            request.requiresOnDeviceRecognition = true
            request.shouldReportPartialResults = false
            let start = ProcessInfo.processInfo.systemUptime
            task = recognizer.recognitionTask(with: request) { [weak self] result, error in
                DispatchQueue.main.async {
                    guard let self, self.active === call else { return }
                    if let result, result.isFinal {
                        let segments = result.bestTranscription.segments.map {
                            ["start": $0.timestamp, "end": $0.timestamp + $0.duration, "text": $0.substring] as [String: Any]
                        }
                        let value: [String: Any] = [
                            "text": result.bestTranscription.formattedString,
                            "language": self.locale(call).identifier.replacingOccurrences(of: "_", with: "-"),
                            "segments": segments,
                            "durationMs": Int((ProcessInfo.processInfo.systemUptime - start) * 1000)
                        ]
                        self.finish(result: value)
                    } else if let error {
                        self.finish(error: error.localizedDescription, code: recognizer.isAvailable ? "failed" : "unavailable")
                    }
                }
            }
            request.append(buffer)
            request.endAudio()
    }
    @objc public func synthesize(_ call: CAPPluginCall) {
        DispatchQueue.main.async { self.runNative(call, synthesis: true) }
    }
    @objc public func cancel(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if let id = call.getString("requestId"), id == self.active?.getString("requestId") {
                if self.nativeEngine != nil { self.cancelWaiters.append(call); self.finish(error: "Speech stopped", code: "cancelled"); return }
                self.finish(error: "Speech stopped", code: "cancelled")
            }
            call.resolve()
        }
    }
    private func finish(result: [String: Any]? = nil, error: String? = nil, code: String = "failed") {
        if let nativeEngine { nativeCancelled = true; gezel_speech_cancel(nativeEngine); return }
        guard let call = active else { return }
        active = nil
        task?.cancel(); task = nil; recognizer = nil
        if let result { call.resolve(result) }
        else { call.reject(error ?? "Speech failed", code) }
    }

    private func runNative(_ call: CAPPluginCall, synthesis: Bool) {
        guard active == nil, UIApplication.shared.applicationState != .background else { call.reject("Speech is busy or Gezel is in the background", "busy"); return }
        guard call.getString("requestId").flatMap(UUID.init(uuidString:)) != nil else { call.reject("A speech request identity is required", "invalid-input"); return }
        if ProcessInfo.processInfo.thermalState == .serious || ProcessInfo.processInfo.thermalState == .critical {
            call.reject("Let this device cool before running speech", "resource-limit"); return
        }
        let model = synthesis ? "kokoro-82m-v1.0" : "whisper-tiny"
        if let requested = call.getString("model"), requested != model { call.reject("The selected speech model is not installed", "unavailable"); return }
        guard ready(synthesis ? "kokoro/model.int8.onnx" : "whisper-tiny.bin") else { call.reject("The offline speech pack is missing from this build", "unavailable"); return }
        let voice = call.getString("voice") ?? "af_heart"
        let index = voices.first { $0["id"] as? String == voice }?["index"] as? Int
        if synthesis && index == nil { call.reject("The selected Kokoro voice is unavailable", "invalid-input"); return }
        // Phoneme ids, produced by the shared @bendyline/gezel/kokoro frontend
        // in the WebView. Nothing native turns text into sound any more, which
        // is how eSpeak NG left the app.
        var tokens: [Int32] = []
        if synthesis {
            let supplied = (call.getArray("tokens") as? [Int]) ?? []
            guard supplied.count >= 3, supplied.count <= 511, supplied.allSatisfy({ $0 >= 0 && $0 <= 177 })
            else { call.reject("Speech phonemes are missing or too long", "invalid-input"); return }
            tokens = supplied.map(Int32.init)
        }
        guard let engine = gezel_speech_create() else { call.reject("Not enough memory for speech", "resource-limit"); return }
        active = call; nativeEngine = engine; nativeCancelled = false
        let root = speechRoot
        worker.async {
            let start = ProcessInfo.processInfo.systemUptime
            var result: [String: Any]?
            var failure: String?
            if synthesis {
                var wav: UnsafeMutablePointer<UInt8>?
                var count = 0
                let speed = Float(min(2, max(0.5, call.getDouble("speed") ?? 1)))
                let code = tokens.withUnsafeBufferPointer { ids in
                    gezel_speech_synthesize(engine, root.appendingPathComponent("kokoro").path,
                                            ids.baseAddress, ids.count, Int32(index!), speed, &wav, &count)
                }
                if code == 0, let wav {
                    let data = Data(bytes: wav, count: count)
                    result = ["wav": data.base64EncodedString(), "meta": ["voice": voice, "model": model, "sampleRate": 24000, "durationSeconds": Double(count - 44) / 48000, "durationMs": Int((ProcessInfo.processInfo.systemUptime - start) * 1000)]]
                } else { failure = String(cString: gezel_speech_error(engine)) }
                gezel_speech_free(wav)
            } else if let encoded = call.getString("audio"), encoded.count <= 5_120_000, let pcm = Data(base64Encoded: encoded), pcm.base64EncodedString() == encoded {
                var text: UnsafeMutablePointer<CChar>?
                let code = pcm.withUnsafeBytes { bytes in
                    gezel_speech_transcribe(engine, root.appendingPathComponent("whisper-tiny.bin").path, bytes.baseAddress?.assumingMemoryBound(to: UInt8.self), bytes.count, call.getString("language"), call.getString("prompt"), &text)
                }
                if code == 0, let text { result = ["text": String(cString: text), "durationMs": Int((ProcessInfo.processInfo.systemUptime - start) * 1000)] }
                else { failure = String(cString: gezel_speech_error(engine)) }
                gezel_speech_free(text)
            } else { failure = "Invalid speech recording" }
            DispatchQueue.main.async {
                gezel_speech_destroy(engine); self.nativeEngine = nil
                let stopped = self.nativeCancelled
                self.finish(result: stopped ? nil : result, error: stopped ? "Speech stopped" : failure, code: stopped ? "cancelled" : "failed")
                let waiters = self.cancelWaiters; self.cancelWaiters = []
                for waiter in waiters { waiter.resolve() }
            }
        }
    }
}
